{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2683: DtypeWarning: Columns (1) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     Id  GTEX-1117F-0426-SM-5EGHI  GTEX-1117F-0626-SM-5N9CS  \\\n",
      "0  ENSG00000000003.10_0                    156.21                   1344.38   \n",
      "1  ENSG00000000003.10_1                     26.85                    181.70   \n",
      "2  ENSG00000000003.10_2                     18.54                    123.89   \n",
      "3  ENSG00000000003.10_3                     35.20                    228.74   \n",
      "4  ENSG00000000003.10_4                     22.08                    162.87   \n",
      "5  ENSG00000000003.10_5                     21.46                    112.96   \n",
      "6  ENSG00000000003.10_6                     33.41                    230.78   \n",
      "7  ENSG00000000003.10_7                     44.25                    220.68   \n",
      "8  ENSG00000000003.10_8                      0.00                      0.00   \n",
      "\n",
      "   GTEX-1117F-0726-SM-5GIEN  GTEX-1117F-1326-SM-5EGHH  \\\n",
      "0                    387.25                   1437.13   \n",
      "1                     51.40                    256.55   \n",
      "2                     32.47                    165.67   \n",
      "3                     60.92                    264.58   \n",
      "4                     37.18                    191.65   \n",
      "5                     34.87                    178.35   \n",
      "6                     57.28                    307.32   \n",
      "7                     54.63                    284.12   \n",
      "8                      0.00                      0.00   \n",
      "\n",
      "   GTEX-1117F-2426-SM-5EGGH  GTEX-1117F-2526-SM-5GZY6  \\\n",
      "0                   1098.50                   1177.90   \n",
      "1                    169.85                    155.72   \n",
      "2                     99.80                     97.64   \n",
      "3                    166.38                    155.09   \n",
      "4                    111.89                     87.91   \n",
      "5                     95.36                     67.63   \n",
      "6                    154.29                    112.08   \n",
      "7                    361.92                     97.03   \n",
      "8                      0.00                      0.00   \n",
      "\n",
      "   GTEX-1117F-2826-SM-5GZXL  GTEX-1117F-2926-SM-5GZYI  \\\n",
      "0                   2158.72                    409.83   \n",
      "1                    368.51                     71.33   \n",
      "2                    231.79                     40.50   \n",
      "3                    403.34                     63.54   \n",
      "4                    260.87                     55.18   \n",
      "5                    191.60                     39.09   \n",
      "6                    388.46                     56.32   \n",
      "7                    299.67                     47.21   \n",
      "8                      0.00                      0.00   \n",
      "\n",
      "   GTEX-1117F-3026-SM-5GZYU  ...   GTEX-ZZPU-2426-SM-5E44I  \\\n",
      "0                   1198.00  ...                    273.17   \n",
      "1                    151.47  ...                     44.29   \n",
      "2                    101.67  ...                     21.50   \n",
      "3                    206.92  ...                     44.67   \n",
      "4                    136.80  ...                     25.58   \n",
      "5                    101.87  ...                     14.55   \n",
      "6                    176.24  ...                     45.18   \n",
      "7                    167.03  ...                     46.05   \n",
      "8                      0.00  ...                      0.00   \n",
      "\n",
      "   GTEX-ZZPU-2626-SM-5E45Y  GTEX-ZZPU-2726-SM-5NQ8O    ensemblGeneId  CHR  \\\n",
      "0                   101.46                  1776.61  ENSG00000000003    X   \n",
      "1                    11.84                   230.68  ENSG00000000003    X   \n",
      "2                    13.67                   163.21  ENSG00000000003    X   \n",
      "3                    19.83                   269.26  ENSG00000000003    X   \n",
      "4                    14.15                   190.12  ENSG00000000003    X   \n",
      "5                    12.65                   150.88  ENSG00000000003    X   \n",
      "6                    18.62                   289.25  ENSG00000000003    X   \n",
      "7                    27.71                   237.91  ENSG00000000003    X   \n",
      "8                     0.00                     0.00  ENSG00000000003    X   \n",
      "\n",
      "      start      stop  strand  exon_count  exon  \n",
      "0  99883667  99884983       -           9     9  \n",
      "1  99885756  99885863       -           9     8  \n",
      "2  99887482  99887565       -           9     7  \n",
      "3  99888402  99888536       -           9     6  \n",
      "4  99888928  99889026       -           9     5  \n",
      "5  99890175  99890249       -           9     4  \n",
      "6  99890555  99890743       -           9     3  \n",
      "7  99891188  99892101       -           9     2  \n",
      "8  99894942  99894988       -           9     1  \n",
      "\n",
      "[9 rows x 8563 columns]\n",
      "Successfully processed ENSG00000000003 at line 10\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "#  import mygene\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "import json\n",
    "import os\n",
    "\n",
    "#  mg = mygene.MyGeneInfo()\n",
    "\n",
    "# WD=\"/hpf/projects/brudno/wangpeiq/sickkids_summer/\"\n",
    "WD=\"../\"\n",
    "SAMPLE_ANNOTATION = WD + \"data/annotation/GTEx_Data_V6_Annotations_SampleAttributesDS.txt\"\n",
    "EXON_EXPR_DATA = WD + \"data/GTEx_Analysis_v6_RNA-seq_RNA-SeQCv1.1.8_exon_reads.txt\"\n",
    "EXON_REF = WD + \"data/gencode.v19.genes.patched_contigs_exons.txt\"\n",
    "PHENOTYPE_ANNOTATION = WD + \"data/annotation/GTEx_Data_V6_Annotations_SubjectPhenotypesDS.txt\"\n",
    "\n",
    "DEST_STORE = WD + \"resources/exon_expr/\"\n",
    "STORAGE_MAPPING = WD + \"resources/mapping\"\n",
    "\n",
    "# GENE_SYMBOL = \"TNMD\"\n",
    "# TISSUES = [\"Muscle -  \", \"Cells - Transformed fibroblasts\"]\n",
    "# MIN_THRESHOLD = 20\n",
    "\n",
    "\n",
    "smp_anno = pd.read_table(SAMPLE_ANNOTATION)\n",
    "smp_anno_extr = smp_anno[[\"SAMPID\", \"SMTSD\"]]\n",
    "\n",
    "# Join exon reference table with exon expression table\n",
    "exon_ref = pd.read_table(EXON_REF)\n",
    "exon_ref[\"Id\"] = exon_ref[\"Gene\"]\n",
    "del exon_ref[\"Gene\"]\n",
    "\n",
    "\n",
    "# Get proper exon number based on strand +/-\n",
    "# Given id = ENSG00000000003.10_0, exon_num = 0\n",
    "# if strand -> + : exon += 1\n",
    "# if strand -> - : exon = exon_count - exon_num\n",
    "def getProperExonNumber(row):\n",
    "    if(row.strand == \"+\"):\n",
    "        return int(row.Id.split(\"_\")[1]) + 1\n",
    "    elif(row.strand == \"-\"):\n",
    "        return row.exon_count - int(row.Id.split(\"_\")[1])\n",
    "    else:\n",
    "        raise Exception(\"invalid strand value (not +/-)\")\n",
    "\n",
    "def processOne(exon, line_num):\n",
    "\n",
    "    gid = exon[\"ensemblGeneId\"].iloc[0]\n",
    "\n",
    "    # exon_count is number of exon per gene; exon indicates proper exon enumeratinon \n",
    "    exon = exon.set_index('Id').join(exon_ref.set_index('Id')).reset_index()\n",
    "    exon[\"exon_count\"] = exon.groupby(\"ensemblGeneId\")[\"ensemblGeneId\"].transform('count')\n",
    "    exon[\"exon\"] = exon.apply(getProperExonNumber, axis=1)\n",
    "    \n",
    "    print(exon)\n",
    "\n",
    "    # dataframe transposed\n",
    "    df_t = exon.transpose()\n",
    "\n",
    "    \n",
    "    # save gene specific metadata to dataframe._metadata\n",
    "    genedata = {}\n",
    "    genedata[\"ensembl_id\"] = exon[\"ensemblGeneId\"].iloc[0]\n",
    "\n",
    "    # Drop irrelevant rows\n",
    "    df_t.columns = df_t.loc[\"exon\"]\n",
    "    df_t = df_t.drop([\"ensemblGeneId\",\n",
    "                      \"CHR\",\n",
    "                      \"start\",\n",
    "                      \"stop\",\n",
    "                      \"strand\",\n",
    "                      \"exon_count\",\n",
    "                      \"exon\",\n",
    "                      \"Id\"])\n",
    "    \n",
    "    \n",
    "    # Group gene-specific exon expression by tissue type\n",
    "    # End result a 53 (SMTSD) x ~10 (exon_counts), each cell holds an array\n",
    "    # merge with sample annotation (specifically SMTSD, tissue type detail)\n",
    "    # Note merging tables has to be done in this step!\n",
    "    df_m = df_t.join(smp_anno_extr.set_index('SAMPID')).set_index(\"SMTSD\")\n",
    "#     print(df_m)\n",
    "\n",
    "#     # populate list by tissue type\n",
    "#     # all_tissue_expr = { exon_num: {tissueType: [ expr...], ...} }\n",
    "#     all_tissue_expr = {}\n",
    "#     for col in df_tissue.columns:\n",
    "#         cur_tissue_expr = df_tissue.groupby(df_tissue.index)[\n",
    "#             col].apply(lambda x: list(x)).to_dict()\n",
    "#         all_tissue_expr[col] = cur_tissue_expr\n",
    "\n",
    "#     genedata[\"exon_expression\"] = all_tissue_expr\n",
    "\n",
    "    \n",
    "#     # separate into 200 bins \n",
    "\n",
    "#     dir_path = os.path.join(DEST_STORE, str(line_num % 200))\n",
    "#     full_path = os.path.join(dir_path, gid)\n",
    "\n",
    "#     if not os.path.exists(dir_path):\n",
    "#         os.makedirs(dir_path)\n",
    "\n",
    "#     with open(full_path, \"w\") as fp:\n",
    "#         json.dump(genedata, fp)\n",
    "#     with open(STORAGE_MAPPING, \"a\") as fp:\n",
    "#         fp.write(gid + '\\t' + full_path + '\\n')\n",
    "    \n",
    "#     del exon\n",
    "#     del df_t\n",
    "#     del df_tissue\n",
    "#     del all_tissue_expr\n",
    "#     del genedata\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Pre-process data for bar plots\n",
    "def subsetBasedOnTissue(data, tissues):\n",
    "    \"\"\" extract expression data for one gene\n",
    "    given a list of tissues in interest \n",
    "        \n",
    "        @param data: pre-computed data (per gene)\n",
    "        @param tissues: [tissueType: str]\n",
    "        @rType:            \n",
    "            [{tissueType: {indexas(exon_num): [expr...]}}...]\n",
    "    \"\"\"\n",
    "    expr = genedata[\"exon_expression\"]\n",
    "    \n",
    "    ret = []\n",
    "    \n",
    "    for tissue in tissues:\n",
    "        cur_tissue_expr = {}\n",
    "        for exon_num in expr:\n",
    "            cur_tissue_expr[exon_num] = expr[exon_num][tissue]\n",
    "        ret.append({tissue: cur_tissue_expr}) \n",
    "    return ret\n",
    "\n",
    "def getDataForPlot(subset):\n",
    "    # plotdata: [ each tissue:{means:[], exons: []}, ... ]\n",
    "    plotdata = []\n",
    "\n",
    "    for t in subset:\n",
    "        tissue_expression = t.values()\n",
    "        exons = []\n",
    "        means = []\n",
    "        stdevs = []\n",
    "        for exons_expr in tissue_expression:\n",
    "            # k: {indexas(exon_num): [expr...]}\n",
    "            for exon_num in exons_expr.keys():\n",
    "                exons.append(exon_num)\n",
    "                means.append(np.mean(exons_expr[exon_num]))\n",
    "                stdevs.append(np.std(exons_expr[exon_num]))\n",
    "        plotdata.append({\n",
    "            \"mean\": means,\n",
    "            \"exon\": exons,\n",
    "            \"stdev\": stdevs\n",
    "        })\n",
    "    return plotdata\n",
    "\n",
    "def plotBarGraph(plotd):\n",
    "    # Plotting barplot y-axis: raw counts, x-axis: exon number for GENE_SYMBOL\n",
    "    # given TISSUES\n",
    "\n",
    "\n",
    "    # These are the \"Tableau 20\" colors as RGB.\n",
    "    tableau20 = [(31, 119, 180), (174, 199, 232), (255, 127, 14), (255, 187, 120),\n",
    "                (44, 160, 44), (152, 223, 138), (214, 39, 40), (255, 152, 150),\n",
    "                (148, 103, 189), (197, 176, 213), (140, 86, 75), (196, 156, 148),\n",
    "                (227, 119, 194), (247, 182, 210), (127, 127, 127), (199, 199, 199),\n",
    "                (188, 189, 34), (219, 219, 141), (23, 190, 207), (158, 218, 229)]\n",
    "    # Scale the RGB values to the [0, 1] range, which is the format matplotlib\n",
    "    # accepts.\n",
    "    for i in range(len(tableau20)):\n",
    "        r, g, b = tableau20[i]\n",
    "        tableau20[i] = (r / 255., g / 255., b / 255.)\n",
    "\n",
    "    num_of_exons = len(plotd[0][\"mean\"])\n",
    "\n",
    "    fig, ax = plt.subplots()\n",
    "    fig.set_size_inches((10, 10))\n",
    "\n",
    "    width = 0.25\n",
    "\n",
    "    ticks = []\n",
    "    ind = np.empty(num_of_exons)\n",
    "\n",
    "    for i, tissue in enumerate(plotd):\n",
    "        ind = np.arange(len(tissue[\"mean\"]))\n",
    "\n",
    "        ax.bar(ind + width * i, tissue[\"mean\"], width,\n",
    "            color=tableau20[i % 20], yerr=tissue[\"stdev\"])\n",
    "        ticks = tissue[\"exon\"]\n",
    "\n",
    "    ax.set_title(\"Raw counts for {} given {}\".format(\n",
    "        GENE_SYMBOL, \",\".join(TISSUES)))\n",
    "    ax.set_ylabel(\"RNAseq exon raw counts\")\n",
    "    ax.set_xlabel(\"Exon number\")\n",
    "\n",
    "    plt.xticks(ind + width / 2, ticks)\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "\n",
    "def getNext(reader):\n",
    "    exon = reader.get_chunk()\n",
    "    exon[\"ensemblGeneId\"] = exon.apply(\n",
    "        lambda row: row[\"Id\"].split(\".\")[0], axis=1)\n",
    "    return exon\n",
    "\n",
    "\n",
    "def iterProcess():\n",
    "\n",
    "    reader = pd.read_table(EXON_EXPR_DATA, chunksize=1)\n",
    "\n",
    "    headers = pd.read_table(EXON_EXPR_DATA, nrows=1).columns\n",
    "    one_gene = pd.DataFrame(columns=headers)\n",
    "\n",
    "\n",
    "    prev_ensembl_id = \"\"\n",
    "    cur_ensembl_id = \"\"\n",
    "\n",
    "    line_num = 1\n",
    "\n",
    "    while True:\n",
    "\n",
    "        cur_row = pd.DataFrame()\n",
    "\n",
    "        try:\n",
    "            cur_row = getNext(reader)\n",
    "        except(StopIteration):\n",
    "            # process the last bit...\n",
    "            processOne(one_gene)\n",
    "            print(\"Successfully processed {} before EOF\".format(prev_ensembl_id))\n",
    "            print(\"stopping iteration...\")\n",
    "            break\n",
    "\n",
    "        cur_ensembl_id = cur_row[\"ensemblGeneId\"].iloc[0]\n",
    "\n",
    "        # Here prev- and cur- ensembl id is correct\n",
    "        if(prev_ensembl_id == \"\" or cur_ensembl_id == prev_ensembl_id):\n",
    "            one_gene = one_gene.append(cur_row, ignore_index=True)\n",
    "        else:\n",
    "            # do processing on exon expression for one gene\n",
    "            # releaase memory before starting on the next gene\n",
    "\n",
    "            processOne(one_gene, line_num)\n",
    "            print(\"Successfully processed {} at line {}\".format(prev_ensembl_id, line_num))\n",
    "            break\n",
    "\n",
    "            del one_gene\n",
    "            one_gene = pd.DataFrame(columns=headers)\n",
    "            one_gene = one_gene.append(cur_row, ignore_index=True)\n",
    "\n",
    "        prev_ensembl_id = cur_ensembl_id\n",
    "        cur_ensembl_id = None\n",
    "        line_num += 1\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    iterProcess()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
